### PR TITLE
PIG 14: Uncertainty estimation

### PR DESCRIPTION
**Pull request by cdeil**
This PR contains some questions and ideas concerning uncertainty estimation / covariance matrix / samples in Gammapy modelling.

I started it to split out the unclear and controversial question of whether to use uncertainties or not from PIG 13 "Gammapy dependencies and distribution" (see #2218).

This PIG is very much related to PIG 7 "Models" (see #1971). That PIG is already very large and has stalled in the past months, and the hope is that by splitting out this aspect of modelling into it's own document and discussion thread, we can manage to finish the design and plan for modelling for Gammapy v1.0 overall more quickly, ideally by July 2019 and most things implemented in Gammapy v0.14 in September 2019.

Anyone with interest or time for this in the coming weeks or months, please consider contributing or taking the lead on this PIG or one of the three sub-sections / questions. For each, we should probably start by more clearly working out what the possible solutions are, and only then after some discussion focus on one proposal and try to work it out at a level of detail that's appropriate for a design document.

@adonath @registerrier @facero @luca-giunti - Maybe you have time to think about the questions here and comment within the near future?

### CODE REVIEW

**Comment by facero** [comment]
Thanks for this first version @cdeil !
I learnt about the existence of astropy.uncertainty and mcerp. Pretty cool tools.

In many ongoing studies (PeVatron or GRBs), people simulate many observation to evaluate errors on flux or sigma. Starting from Npred and assuming a Poisson distribution could we somehow use astropy.uncertainty to estimate the errors without having to run the same observations > 100 times ?

Concerning MCMC, at the last coding sprint we had quick look with Axel and it didn't seem that much complicated to implement it in gammapy fitting. 
I'm not sure to be able to come to the next sprint but I could spend a day or two to work on this from Saclay.

**Comment by cdeil** [comment]
> In many ongoing studies (PeVatron or GRBs), people simulate many observation to evaluate errors on flux or sigma. Starting from Npred and assuming a Poisson distribution could we somehow use astropy.uncertainty to estimate the errors without having to run the same observations > 100 times ?

I'm not sure. I think for some things one can get away with just doing one "simulation" with `n_obs = n_pred", for some it's sufficient to do one real "simulation" with `n_obs = np.random.poisson(n_pred)` and for some things one has to run many simulations and look at distributions. E.g. for sensitivity computations I was wondering that as well previously.

> Concerning MCMC, at the last coding sprint we had quick look with Axel and it didn't seem that much complicated to implement it in gammapy fitting.
I'm not sure to be able to come to the next sprint but I could spend a day or two to work on this from Saclay.

@facero or @adonath - If you  have an idea what to add to `gammapy.utils.fitting`, could you please briefly describe it here? Preferably directly by editing the relevant PIG section and putting some text in the document, alternatively you could put your suggestion in a comment field or a gist and I'd put it in the document.

**Comment by adonath** [comment]
@facero and @cdeil Technically it is easy to add the emcee backend, but the related changes to the Parameter API or which new APIs we have to introduce is less clear. I think the main questions we have to address are:

- How to store / define parameter priors?
- How to store sampled parameter values / distributions?

This will require a (written) proposal and subsequent discussion. I agree it could be in scope for this PIG to propose something. I won't have time to work on this up to the Gammapy coding sprint, but I'm happy to review / discuss any proposal.

**Comment by QRemy** [comment]
For 1D spectral fitting emcee would be fine, but for 3D fitting of multiple sources+backgrounds
we should propose other MC methods based on nested sampling that are more efficient for high dimensionality problems.  There is several solutions available in python: [Multinest](https://arxiv.org/pdf/0809.3437.pdf), [Polychord](https://arxiv.org/pdf/1506.00171.pdf), [Dynesty](https://arxiv.org/pdf/1904.02180.pdf).  I can have a look into that.

**Comment by cdeil** [comment]
@QRemy - Thanks for getting involved here!

Agreed that ``emcee`` shouldn't be the only / likely won't be the long-term solution.
My understanding is that Hamiltonian MC with gradient-based sampling is the state of the art (https://docs.pymc.io/notebooks/getting_started.html), and Tensorflow Probability is getting a lot of traction (see https://github.com/pymc-devs/pymc4).

The problem is that those things are "all-in" modeling frameworks, one cannot use them like ``emcee`` with a minimal parameter array interface in the likelihood function. I'm not familiar with the other packages that you mention. Do they allow for such an interface and can be used as optional "backends" in addition to emcee, or are they of the "all-in" kind that would require all our models to be written in their framework?

Please note that the scope of this PIG should be to describe work that we can achieve by ~ September 2019, i.e. very short-term. Maybe it would be best to try and move the outside helper functions from
https://docs.gammapy.org/0.12/notebooks/mcmc_sampling.html?highlight=emcee#Estimate-parameter-correlations-with-MCMC into ``gammapy.utils.fitting`` and to have some built-in support that is usable and tested, and then post v1.0 aim for a big rewrite of ``gammapy.utils.fitting`` to support much more?

**Comment by facero** [comment]
I didn't know about those tools. Thanks for the link. The link in the pdf is dead but I managed to track some of them to their githubs [PolyChordLite](https://github.com/PolyChord/PolyChordLite) and [MultiNest](https://github.com/farhanferoz/MultiNest).
Apparently the PolyChord folks have now started a company for BigData stuff and have a paid and one free version for academic use.
Looking quickly you, both seems to have a Python wraper but don't how easy they are to use.
Like this [PyMultiNest](https://github.com/JohannesBuchner/PyMultiNest).

I guess for gammapy, it could be better to provide a simple tool like emcee at first and maybe a notebook to explain how to use more adavnced tools.
I'm not familiar at all with those tools but having a deeper look at these would be nice. I'll be happy to learn about the difference between  MultiNest and simple MCMC.
What is this concept behind Nested modeling ?

**Comment by cdeil** [comment]
@QRemy @facero - Please note that we have to focus on v1.0 and e.g. implementing a first version of model serialisation is much more important than fancy sampling methods. The discussion / work on sampling should either be to improve a little bit by moving the helper functions from https://docs.gammapy.org/0.12/notebooks/mcmc_sampling.html?highlight=emcee#Estimate-parameter-correlations-with-MCMC into our modeling framework in `gammapy.utils.fitting`, or if that is difficult for some reason, do nothing on this now and postpone that work to post v1.0.

I brought it up because I thought that we can with limited effort (a few days of discussion and experimentation, followed by a few weeks of work by one person on code, tests, docs) introduce samples into some place in `gammapy.utils.fitting`, coming either from MCMC or from multivariate normal sampling of the covariance matrix (see the other sections), providing a general and better solution for error propagation.

**Comment by facero** [comment]
Agreed.
I plan to take some time to implement the helper functions from my notebook into gammapy.utils.fitting the week of the coding sprint. I won't be able to join you but I'll be in touch via Slack.

**Comment by QRemy** [comment]
Nested sampling decompose the posterior distribution into subsets simpler to sample, instead of trying to do it directly.
@facero @cdeil - For Multinest and Dynesty you can give a log-likelihood function and a prior function in a similar way than emcee. Here you have the same example for [dynesty](https://github.com/joshspeagle/dynesty/blob/master/demos/Examples%20--%20Linear%20Regression.ipynb) and [emcee](https://emcee.readthedocs.io/en/v2.2.1/user/line/), and another one for [multinest](https://github.com/JohannesBuchner/PyMultiNest/blob/master/pymultinest_demo.py).

**Comment by cdeil** [comment]
@QRemy - then we could try to add the MCMC things (posterior, samples, interface to samples) in a generic and uniform way, and have an extensible registry of samplers in `gammapy.utils.fitting`, similar to what we have for optimisers here:
https://github.com/gammapy/gammapy/blob/369b3359e9b5bc580a855f76efc0e6b8bb733e04/gammapy/utils/fitting/fit.py#L20-L47
Not sure how it would look, i.e. if we add this to this one registry, or have a second one.
Practically it's probably still the most efficient to just integrate posterior, samples and emcee interface into gammapy.utils.fitting, and then to generalise and add a second one and expose them as optional "backends" users can choose when calling `sample` (or would it also be "optimise"? - that's the parts we need a proposal for, thus PRs to fill some infos in this PIG welcome any time!)

**Comment by cdeil** [comment]
The `evaluate_error` using `uncertainties` broke with Numpy 1.17 which was released a week ago.
- https://dev.azure.com/gammapy/gammapy/_build/results?buildId=1497
- https://github.com/numpy/numpy/releases/tag/v1.17.0

I think it'll be fixed soon, so for now it's OK, no urgent changes needed in Gammapy:
https://github.com/lebigot/uncertainties/issues/89

**Comment by cdeil** [comment]
Just wanted to leave a note here since I came across:
https://www.astro.rug.nl/software/kapteyn/kmpfittutorial.html#confidence-and-prediction-intervals

That's roughly what is needed for generic error propagation:
https://en.wikipedia.org/wiki/Propagation_of_uncertainty

We have the parameter covariance matrix from the fit. So all one needs to compute is how the quantity of interest changes when each free parameter is changed. This can be done as a numerical derivative in a generic way. Then do the matrix product.

The nice thing about that is that it works for everything: currently we only have `dnde` for some spectral model. That will work for `dnde` or `flux` or any other derived quantity from any spectral or spatial or any model.

The other solution outlined in the PIG, to [MultiNorm.sample](https://multinorm.readthedocs.io/en/latest/api.html#multinorm.MultiNorm.sample) parameters, compute the quantity of interest, then if one wants a value and error instead of a PDF use [MultiNorm.from_samples](https://multinorm.readthedocs.io/en/latest/api.html#multinorm.MultiNorm.from_samples) is also only ~ 20 lines of code and very generic.

I really think we should give this a shot, because the gains are large for users (many need error estimates on derived quantities, not just fit parameters) and for us (can get rid of `uncertainties` and special code in each derived quantity computation, e.g. `dnde`, `flux`, ...). It is a complex task (especially if we use numerical derivatives step sizes will be an issue, and if we don't manage to use Numpy broadcasting, efficiency could be an issue), but maybe a prototype could be written in a day or two in a notebook, and then if it can reproduce a `dnde` error band, we spend the ~ week to rework our `uncertainties` code.

Anyone interested?

**Comment by cdeil** [comment]
Here's a prototype function that we could put a `SpectralModel.evaluate_error` in the base class:
https://nbviewer.jupyter.org/gist/cdeil/9b73a0c712fd23742a9ff911b5a7fab4#Differentials

It works, but there's a few important questions:
- How to vectorise the code to work with `energy` arrays? (I tried for 3 min and failed, could try again tomorrow, should be possible with a few small changes, using Numpy broadcasting)
- We want `integrate_error` as well? How to structure the code to avoid code duplication?
- Is this stepsize of 1% parameter error OK? Or is there a better way?
- Should we create a new model when stepping in a parameter? Or stick with the current step back and forth, changing temporarily the model state?

For the MCMC sampling technique, I didn't prototype it yet. But it's already clear that there's the important question whether this needs a Python for loop, or whether we can get model evaluation to work for parameter arrays, and if yes, how to implement this. Note that Astropy has this implemented already, a feature called [model sets](https://docs.astropy.org/en/stable/modeling/#model-sets), and I think they do it by having quite a bit of Numpy broadcasting logic in `__call__`, allowing `evaluate` to be written in a simple way as if scalars come in. Maybe just do the Python for loop for now, and leave a comment in the PIG that we might consider model sets for Gammapy in the future, but it's not urgently needed for now?

@adonath or anyone interested to collaborate on this - Thoughts?

**Comment by cdeil** [comment]
Here's one way to implement `dnde_error` and `integral_error` without code duplication:
https://nbviewer.jupyter.org/gist/cdeil/27383e9d13125b9ba0c7784d620eda6f#Differentials
Is this the way to go? Or is there a better way?

Here's a prototype for the sample-based error propagation:
https://nbviewer.jupyter.org/gist/cdeil/27383e9d13125b9ba0c7784d620eda6f#Samples
It basically works, but `np.std(samples)` is always smaller than the differential-based error propagation, even if I set a small `err_scale=0.01`. I don't understand, it should come out the same (apart from sampling noise), no?
Can anyone spot the bug, or explain why it's not the same?

**Comment by cdeil** [comment]
Draft: https://github.com/gammapy/gammapy/blob/562cb870e16e0e59b3cd8ddf20fcbea26f140acf/docs/development/pigs/pig-014.rst

@adonath @QRemy - I've added you as co-authors, please review and comment.
@facero - Do you want to co-author as well? (i.e. basically agree with the proposal)

I think we should try to go to official review with this PIG very soon, ideally in the next 1-2 days.
If we give next Monday as deadline for comments, we can aim to approve and implement the PIG next week.

If we don't do that, but try to work out the details of https://nbviewer.jupyter.org/gist/cdeil/27383e9d13125b9ba0c7784d620eda6f and the API for samples, we will be in trouble, because either we make no progress and implement this in the coming weeks and for v0.15, or we do, but then circulating the PIG and asking for feedback becomes useless.

That said, of course it would be nice to improve the description and prototype notebook as much as possible before going to wider review, so please help out via coding or comments.

File: docs/development/pigs/pig-014.rst [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+- `Error estimation in astronomy: A guide`_
+- `Frequentism and Bayesianism: A Python-driven Primer`_
+
+In Gammapy we currently use the `uncertainties`_ package to propagate errors to
+differential fluxes ``spectral_model.evaluate_error`` and integral fluxes
+``spectral_model.integral_error``. It is using autodiff to compute differentials
+of derived quantities wrt. model parameters, which is accurate and fast where it
+works. However, ``uncertainties`` doesn't support ``astropy.units.Quantity``,
+making the current spectral model flux evaluation very convoluted, and it
+doesn't work at all for some spectral models, which has been a frequent
```
Comment (registerrier) [inline docs/development/pigs/pig-014.rst]: you could explicitly state that is fails for non analytical models, which is the case for most of physical models (e.g. EBL absorption or naima models)

**Comment by cdeil** [comment]
@adonath - Do you know how to solve the energy broadcasting question? Or should we leave this to be figured out next week, assuming we can? That's the one concrete thing needed to actually make the proposed change.

**Comment by adonath** [comment]
@cdeil Yes, I'm currently working on it and have a solution. I can update the prototype notebook, that you've linked (I guess I can checkout gists and push to it?).

**Comment by cdeil** [comment]
@adonath - Awesome!

Can you please put the notebook here?
https://github.com/gammapy/gammapy-extra/tree/master/experiments

I can then update the PIG document, and also implement the comments by @registerrier and then we can circulate. Or if you want to do these edits directly, go ahead.

**Comment by cdeil** [comment]
I think we should put Monday, Nov 18 as deadline for comments, and put a 5+5 min presentation on this for https://github.com/gammapy/gammapy-meetings/tree/master/coding-sprints/2019-11_Granada#monday
It's likely that this will yield good feedback, whereas email often does not.
Then we merge in the PIG Tuesday morning, and implement it that week.

**Comment by facero** [comment]
Thanks @cdeil for this PIG.

I just read PIG 14 in details and not being fully familar with `SpectralModel.evaluate_error` could you give a bit more intro (could be just 1-2 lines and or the wikipedia link) and 1 example of its usage. It's not the error on the parameters but the errors on the spectra (to produce the butterfly plot ) right ?
It will help people understand the scope of this PIG.
Also if you say in a few sentences what is the  "differential based error propagation" that would help. I understand that PIGs are not documentation but it helps people to comment and can be the base for futur doc as well.
Would that be able to propagate errors for models like EBL ?

I also like the samples approach if this is something astropy provides it one less dependency and naturally support Quantities. if the model evaluation is slow this could indeed be pretty slow but should be doable to parrallelize.
Could the samples provide asymmetric errors ?

I support the use of autograd for a futur version, it is being used by another project I'm working on and it seems quite powerful. The other alternative, JAX seems designed more for GPUs and need some extra dependencies and is not supported for Windows.

**Comment by cdeil** [comment]
> It's not the error on the parameters but the errors on the spectra (to produce the butterfly plot ) right ?

Yes. I'll make the text clearer.

> Would that be able to propagate errors for models like EBL ?

Yes. I'll mention it explicitly.

> Could the samples provide asymmetric errors ?

Yes. With samples you get the full distribution, and then you use `mean` and `std` if you want to get the mean and standard error.

Did you see the example?
https://nbviewer.jupyter.org/gist/cdeil/27383e9d13125b9ba0c7784d620eda6f#Samples

I don't understand why the MC sample error can't reproduce the `3.6e-12` value!?

I think it should, since even if the methods are different computationally, both measure the local likelihood curvature in a small region around the best-fit values, they should yield the same parameter error within sampling and numerical derivative noise, no?

There's something fishy going on ...

**Comment by facero** [comment]
I downloaded and executed the notebook.
I played with the parameters for the sampling and even with a large number of samples I could not reproduce something close to `3.6e-12`. I get something approaching `2.98e-12` with a very high number of samples.
Not sure why but when I set `err_scale=10` I get :
=> `3.761030834315111e-11 3.425905842340224e-12`
closer to the results obtained with `uncertainties` 

What is the reason of this `err_scale` param ? it seems to have a strong impact of the estimated error ;-(

**Comment by adonath** [comment]
@cdeil Here is the updated prototype https://github.com/gammapy/gammapy-extra/blob/master/experiments/uncertainty_estimation_prototype.ipynb. Shall I update the link in the PIG>

**Comment by cdeil** [comment]
@adonath - Thanks!

I've updated the notebook in https://github.com/gammapy/gammapy-extra/blob/master/experiments/uncertainty_estimation_prototype.ipynb - we need to discuss a bit how exactly we want to implement this, i.e. whether to go through `evaluate` or `__call__`. I think some the models with data like e.g. EBL absorption require to go through `__call__`, so the approach in `gradient_scipy` might not work there?

I've also done a small update to the PIG in a6f1e4aaac05d180c525b8a026b9fe5d7646ac7b to link to the notebook, and to clarify the questions raised by @facero and @registerrier above. Going for official review now with deadline Nov 18.

**Comment by adonath** [comment]
@cdeil The `gradient_scipy` was my first try to get a version that supports broadcasting and multiple energies (because I thought I would get it for free using `approx_jacobian`). I think it should also work in the cases, where `.evaluate` is an instance method (`TemplateSpectralModel` etc.). But it lacks proper support for units and has no possibility to adjust the `eps` per parameter.  That's why I also adapted your method, that goes through `__call__` (and maybe we should just use this one...).

I think the only important point is to split out the gradient computation from `.evaluate_error`, so that it can be used in future for fitting as well, or be replaced by using other 3rd party packages such as `numdifftools` or `autograd` in future.

**Comment by cdeil** [comment]
> it lacks proper support for units 

@adonath - That prototyping brings up major questions for the model evaluation, which we probably should discuss in #2443 and find a solution for very soon.

Gradients and covariance matrices have mixed units, so generally one should use Numpy arrays, not quantities. So we should change all models and tests in Gammapy to use Numpy arrays, not quantity? Or have a two-layer approach like with `__call__` and `evaluate` where `__call__` uses Quantity (and SkyCoord and Time objects), but then `evaluate` only uses Numpy arrays? What about e.g. `spectral_model.integral` or `spatial_model.evaluate_geom`, what types of inputs should they support? Do we aim for a full solution where model evaluation only works with Python floats and Numpy arrays, while still offering a Quantity, SkyCoord, Time based interface for end users, but internally going through the faster route where units are fixed?

**Pull request merged**